### ISSUE TITLE
How do you correctly restart training? 

### ISSUE THREAD

**Issue by vsumaria** [ulissigroup/amptorch#83]
I am trying to restart training using pre-trained weights from the checkpoint. I load then using the load_pretrained function, but the training seems to start from the beginning. What am I doing wrong ?

**Comment by zulissi** [comment]
You're looking for either a skorch regressor with `warm_start=True` or using `partial_fit` instead of `fit`. As written in `train` it always re-initializes.

This is a good suggestion; maybe we should have an option to set warm_start in the skorch model or as another config to pass around?

**Comment by mshuaibii** [comment]
@zulissi This shouldn't be an issue. The skorch regressor loads the weights, optimizer, history after it is reinitialized: https://github.com/ulissigroup/amptorch/blob/9b71f1f2d2b5a49dbd2c705a97d6b6f0a9ce3bb8/amptorch/trainer.py#L303-L318. Additionally, we have tests in place to test this specific functionality: https://github.com/ulissigroup/amptorch/blob/master/amptorch/tests/pretrained_test.py.

I ran a quick test using the `train_example.py` we have:

No retraining:
![image](https://user-images.githubusercontent.com/45150244/107850507-ece52f00-6dd0-11eb-87f1-abc29fa4422c.png)

Retraining:
![image](https://user-images.githubusercontent.com/45150244/107850519-025a5900-6dd1-11eb-8f9f-b967b16aa606.png)

We see that the losses are indeed starting from a point close to the end of the previous run.

@vsumaria Can you share how you're attempting to retrain. Maybe the tests linked above can help provide guidance.

All I've done to generate the above images was:
```
trainer = AtomsTrainer(config) % using the same config used earlier
trainer.load_pretrained(checkpoint_path
trainer.train()
```

**Comment by zulissi** [comment]
Thanks @mshuaibii!

**Comment by vsumaria** [comment]
I am doing exactly that, loading the AtomsTrainer with same config, then loading the pretrained model using the checkpoint_path and trainer.train(). But my last iteration in the previous training with decreasing loss is this: 

<img width="940" alt="Screen Shot 2021-02-13 at 10 43 31 AM" src="https://user-images.githubusercontent.com/33334729/107858280-5290e580-6de8-11eb-9493-aa4f900e5e18.png">

And the one after I restart is this:
<img width="919" alt="Screen Shot 2021-02-13 at 10 45 52 AM" src="https://user-images.githubusercontent.com/33334729/107858372-a6033380-6de8-11eb-8593-25db4d467b5f.png">

That's what is confusing me.

**Comment by vsumaria** [comment]
My learning rate also resets if I am using a LR scheduler.

**Comment by vsumaria** [comment]
> You're looking for either a skorch regressor with `warm_start=True` or using `partial_fit` instead of `fit`. As written in `train` it always re-initializes.
> 
> This is a good suggestion; maybe we should have an option to set warm_start in the skorch model or as another config to pass around?

I tried "warm_start= True", it starts correctly, but you see the loss function starting to rise and then decrease. I am not 100% sure how the gradients get initialized when using this setting. But it at least not restarting the whole training.

**Comment by mshuaibii** [comment]
> I am doing exactly that, loading the AtomsTrainer with same config, then loading the pretrained model using the checkpoint_path and trainer.train(). But my last iteration in the previous training with decreasing loss is this:
> 
> <img alt="Screen Shot 2021-02-13 at 10 43 31 AM" width="940" src="https://user-images.githubusercontent.com/33334729/107858280-5290e580-6de8-11eb-9493-aa4f900e5e18.png">
> 
> And the one after I restart is this:
> <img alt="Screen Shot 2021-02-13 at 10 45 52 AM" width="919" src="https://user-images.githubusercontent.com/33334729/107858372-a6033380-6de8-11eb-8593-25db4d467b5f.png">
> 
> That's what is confusing me.

Yeah this is strange...I'm unable to replicate the `warm_start` issue. Both give me the same results which is expected given that we manually initialize/override settings when we call `load_pretrained`. Can you try these tests out on the `trainer_example.py` and see if you get similar issues, this may make it easier to debug.

> My learning rate also resets if I am using a LR scheduler.

Yeah currently the learning rate scheduler reinitializes after training. Skorch doesn't save `lr` information in the history which makes this a little tricky: https://github.com/skorch-dev/skorch/issues/739. I'll take a look and see if I can find a work around.

**Comment by zulissi** [comment]
> I tried "warm_start= True", it starts correctly, but you see the loss function starting to rise and then decrease. I am not 100% sure how the gradients get initialized when using this setting. But it at least not restarting the whole training.

This sounds like the learning rate got reset to the original learning rate, so the loss goes up then goes back down as learning rate drops.

**Comment by vsumaria** [comment]
> > I tried "warm_start= True", it starts correctly, but you see the loss function starting to rise and then decrease. I am not 100% sure how the gradients get initialized when using this setting. But it at least not restarting the whole training.
> 
> This sounds like the learning rate got reset to the original learning rate, so the loss goes up then goes back down as learning rate drops.

Oh, yes. You are right. I didn't notice it then. Not sure why the LR increased though. I'll try and understand that. 



> > I am doing exactly that, loading the AtomsTrainer with same config, then loading the pretrained model using the checkpoint_path and trainer.train(). But my last iteration in the previous training with decreasing loss is this:
> > <img alt="Screen Shot 2021-02-13 at 10 43 31 AM" width="940" src="https://user-images.githubusercontent.com/33334729/107858280-5290e580-6de8-11eb-9493-aa4f900e5e18.png">
> > And the one after I restart is this:
> > <img alt="Screen Shot 2021-02-13 at 10 45 52 AM" width="919" src="https://user-images.githubusercontent.com/33334729/107858372-a6033380-6de8-11eb-8593-25db4d467b5f.png">
> > That's what is confusing me.
> 
> Yeah this is strange...I'm unable to replicate the `warm_start` issue. Both give me the same results which is expected given that we manually initialize/override settings when we call `load_pretrained`. Can you try these tests out on the `trainer_example.py` and see if you get similar issues, this may make it easier to debug. Side note - the checkpoint that get's stored is marked by a `+` in the `cp` column of the output. For instance, line 934 of your output wasn't checkpointed so we shouldn't be comparing the retrained numbers to that value but rather the last `+`. Given how high that number is, I suspect the last checkpoint was still much lower than the retrained 1st epoch, but maybe you can confirm.
> 
> > My learning rate also resets if I am using a LR scheduler.
> 
> Yeah currently the learning rate scheduler reinitializes after training. Skorch doesn't save `lr` information in the history which makes this a little tricky: [skorch-dev/skorch#739](https://github.com/skorch-dev/skorch/issues/739). I'll take a look and see if I can find a work around.

Ok. I didn't realize the "+" indicated the checkpoint. But yes, the checkpoint is changed and is not the the 1st epoch for sure. 
Btw, How exactly do you decide which epoch is saved as the checkpoint ?

**Comment by vsumaria** [comment]
@mshuaibii you are right. I tried to restart training, it is behaving correctly. I didn't realize the "+" sign meant writing of new check points. (AMP generated checkpoints at equal epoch intervals). 

Btw, how do you decide which epoch should be used as the checkpoint ?

Sorry for the confusion. I think the LR reseting also might have been because of the same reason.

**Comment by zulissi** [comment]
It's a skorch callback for the best loss https://github.com/ulissigroup/amptorch/blob/b18e3cb56b8c791454a769e89e4223eee1c8707d/amptorch/metrics.py#L129

For larger datasets saving the checkpoint with the lowest validation error is probably better, but most of our tests have been with such small datasets that any train/validation split isn't very helpful.

**Comment by vsumaria** [comment]
I think that's a good rule of thumb to also avoid overfitting the potential. 

Closing this comment, I think the retraining is working fine, just didn't realize what check point was being used for the retraining. (I'd update the main README file which is still using the "load_checkpoint" function @mshuaibii)
